crates/storage/provider/src/providers/blockchain_provider.rs

```diff
@@ -1148,26 +1147,19 @@ impl<N: ProviderNodeTypes> StateProviderFactory for BlockchainProvider2<N> {
     fn pending(&self) -> ProviderResult<StateProviderBox> {
         trace!(target: "providers::blockchain", "Getting provider for pending state");
 
-        if let Some(block) = self.canonical_in_memory_state.pending_block_num_hash() {
-            let historical = self.database.history_by_block_hash(block.hash)?;
-            let pending_provider =
-                self.canonical_in_memory_state.state_provider(block.hash, historical);
-
-            return Ok(Box::new(pending_provider));
+        if let Some(pending) = self.canonical_in_memory_state.pending_state() {
+            // we have a pending block
+            return Ok(Box::new(self.block_state_provider(&pending)?));
         }
 
         // fallback to latest state if the pending block is not available
         self.latest()
     }
 
     fn pending_state_by_hash(&self, block_hash: B256) -> ProviderResult<Option<StateProviderBox>> {
-        let historical = self.database.history_by_block_hash(block_hash)?;
-        if let Some(block) = self.canonical_in_memory_state.pending_block_num_hash() {
-            if block.hash == block_hash {
-                let pending_provider =
-                    self.canonical_in_memory_state.state_provider(block_hash, historical);
-
-                return Ok(Some(Box::new(pending_provider)))
+        if let Some(pending) = self.canonical_in_memory_state.pending_state() {
+            if pending.anchor().hash == block_hash {
+                return Ok(Some(Box::new(self.block_state_provider(&pending)?)));
```

Comment:
why are we checking that the anchor is equal here, instead of the `pending.hash()`?

Reply:
ah good point, 
bad autocomplete